File: array_of_products/array_of_products.c
#include <stdio.h>
#include <stdlib.h>

int* arrayOfProducts(int* array, int length) {
    int* products = (int*)malloc(length * sizeof(int));

    for (int i = 0; i < length; i++) {
        products[i] = 1;
    }

    int product = 1;
    for (int i = length - 1; i >= 0; i--) {
        products[i] = product;
        product *= array[i];
    }

    product = 1;
    for (int i = 0; i < length; i++) {
        products[i] *= product;
        product *= array[i];
    }

    return products;
}

int main() {
    int array[] = {5, 1, 4, 2};
    int length = sizeof(array) / sizeof(array[0]);

    int* result = arrayOfProducts(array, length);

    for (int i = 0; i < length; i++) {
        printf("%d ", result[i]);
    }
    printf("\n");

    free(result);
    return 0;
}

File: array_of_products/array_of_products.rs
fn array_of_products(array: &[i32]) -> Vec<i32> {
    let length = array.len();
    let mut products = vec![1; length];

    let mut product = 1;
    for i in (0..length).rev() {
        products[i] = product;
        product *= array[i];
    }

    product = 1;
    for i in 0..length {
        products[i] *= product;
        product *= array[i];
    }

    products
}

fn main() {
    let array = vec![5, 1, 4, 2];
    let result = array_of_products(&array);
    println!("{:?}", result);
}
